cranelift/codegen/src/isa/x64/lower.rs

```diff
@@ -1732,6 +1748,148 @@ fn lower_insn_to_regs<C: LowerCtx<I = Inst>>(
             });
         }
 
+        Opcode::AtomicRmw => {
+            // This is a simple, general-case atomic update, based on a loop involving
+            // `cmpxchg`.  Note that we could do much better than this in the case where the old
+            // value at the location (that is to say, the SSA `Value` computed by this CLIF
+            // insn) is not required.  In that case, we could instead implement this using a
+            // single `lock`-prefixed x64 read-modify-write instruction.  Also, even in the case
+            // where the old value is required, for the `add` and `sub` cases, we can use the
+            // single instruction `lock xadd`.  However, those improvements have been left for
+            // another day.
+            let r_dst = output_to_reg(ctx, outputs[0]);
+            let mut r_addr = input_to_reg(ctx, inputs[0]);
+            let mut r_arg2 = input_to_reg(ctx, inputs[1]);
+            let ty_access = ty.unwrap();
+            assert!(is_valid_atomic_transaction_ty(ty_access));
+            let memflags = ctx.memflags(insn).expect("memory flags");
+            let srcloc = if !memflags.notrap() {
+                Some(ctx.srcloc(insn))
+            } else {
+                None
+            };
+            // Make sure that both args are in virtual regs, since in effect we have to do a
+            // parallel copy to get them safely to the AtomicRMW input regs, and that's not
+            // guaranteed safe if either is in a real reg.
+            r_addr = ctx.ensure_in_vreg(r_addr, types::I64);
+            r_arg2 = ctx.ensure_in_vreg(r_arg2, types::I64);
+            // Move the args to the preordained AtomicRMW input regs.  Note that `AtomicRMW`
+            // operates at whatever width is specified by `ty`, so there's no need to
+            // zero-extend `r_arg2` in the case of `ty` being I8/I16/I32.
+            ctx.emit(Inst::gen_move(
+                Writable::from_reg(regs::r9()),
+                r_addr,
+                types::I64,
+            ));
+            ctx.emit(Inst::gen_move(
+                Writable::from_reg(regs::r10()),
+                r_arg2,
+                types::I64,
+            ));
+            // Now the AtomicRMW (pseudo-) insn itself
+            let op = AtomicRMWOp::from(inst_atomic_rmw_op(ctx.data(insn)).unwrap());
+            ctx.emit(Inst::AtomicRMW {
+                ty: ty_access,
+                op,
+                srcloc,
+            });
+            // And finally, copy the preordained AtomicRMW output reg to its destination.
+            ctx.emit(Inst::gen_move(r_dst, regs::rax(), types::I64));
+            // Also, %r11 is trashed.  `fn x64_get_regs` must mention that.
```

Comment:
Could we remove this comment? It's a bit weird to read this here, since there's a sequence of vcode insts, and the vcode inst actually trashing this register already mentions it does that by the code in `get_regs`.